Microsoft.O365.Security.Native.ETW/Callbacks.hpp

```diff
@@ -0,0 +1,154 @@
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#pragma once
+
+#include <krabs.hpp>
+
+#include "EventRecord.hpp"
+#include "EventRecordError.hpp"
+
+using namespace System;
+using namespace System::Runtime::InteropServices;
+
+namespace Microsoft { namespace O365 { namespace Security { namespace ETW {
+
+    /// <summary>
+    /// Delegate called when a new ETW <see cref="O365::Security::ETW::EventRecord"/> is received.
+    /// </summary>
+    public delegate void IEventRecordDelegate(IEventRecord^ record);
+
+    /// <summary>
+    /// Delegate called when a new ETW <see cref="O365::Security::ETW::EventRecordMetadata"/> is received.
+    /// </summary>
+    public delegate void IEventRecordMetadataDelegate(IEventRecordMetadata^ record);
+
+    /// <summary>
+    /// Delegate called on errors when processing an <see cref="O365::Security::ETW::EventRecord"/>.
+    /// </summary>
+    public delegate void EventRecordErrorDelegate(IEventRecordError^ error);
+
+    delegate void EventNativeDelegate(const EVENT_RECORD&, const krabs::trace_context&);
+
+    delegate void EventErrorNativeDelegate(const EVENT_RECORD&, const std::string&);
+
+    ref class CallbackBridge {
+    private:
+
+        EventNativeDelegate^ eventDelegateKeepAlive_;
+        EventErrorNativeDelegate^ errorDelegateKeepAlive_;
+        EventRecordMetadata^ metadata_;
+        EventRecord^ record_;
+        EventRecordError^ error_;
+
+        EventRecordMetadata^ CreateMetadata(const EVENT_RECORD& record);
+        EventRecord^ CreateRecord(const EVENT_RECORD& record, const krabs::schema& schema, krabs::parser& parser);
+        EventRecordError^ CreateError(const std::string& error_message, const EVENT_RECORD& record);
+
+        void EventNotification(const EVENT_RECORD& record, const krabs::trace_context& trace_context);
+        void ErrorNotification(const EVENT_RECORD& record, const std::string& error_message);
+
+    public:
+        IEventRecordMetadataDelegate^ OnMetadata;
+        IEventRecordDelegate^ OnEvent;
+        EventRecordErrorDelegate^ OnError;
+
+        krabs::c_provider_callback GetOnEventBridge()
+        {
+            if (!eventDelegateKeepAlive_)
+                eventDelegateKeepAlive_ = gcnew EventNativeDelegate(this, &CallbackBridge::EventNotification);
+            return (krabs::c_provider_callback)Marshal::GetFunctionPointerForDelegate(eventDelegateKeepAlive_).ToPointer();
+        }
+
+        krabs::c_provider_error_callback GetOnErrorBridge()
+        {
+            if (!errorDelegateKeepAlive_)
+                errorDelegateKeepAlive_ = gcnew EventErrorNativeDelegate(this, &CallbackBridge::ErrorNotification);
+            return (krabs::c_provider_error_callback)Marshal::GetFunctionPointerForDelegate(errorDelegateKeepAlive_).ToPointer();
+        }
+    };
+
+    EventRecordMetadata^ CallbackBridge::CreateMetadata(const EVENT_RECORD& record)
```

Comment:
nit: I don't like "Create" for the verb here because the instances are reused. I think I'd be less concerned if they weren't public. Maybe "Wrap" or maybe even "Update" would be better?

Reply:
Fixed